File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/support/RequeryPersistableEntityInformation.java
package com.coupang.springframework.data.requery.repository.support;

import io.requery.meta.EntityModel;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.domain.Persistable;

/**
 * Extension of {@link RequeryEntityModelEntityInformation} that consideres methods of {@link Persistable} to lookup the id.
 *
 * @author debop@coupang.com
 * @since 18. 6. 7
 */
@Slf4j
public class RequeryPersistableEntityInformation<T extends Persistable<ID>, ID>
    extends RequeryEntityModelEntityInformation<T, ID> {

    public RequeryPersistableEntityInformation(@NotNull Class<T> domainClass,
                                               @NotNull EntityModel entityModel) {
        super(domainClass, entityModel);
    }

    @Override
    public boolean isNew(@NotNull T entity) {
        return entity.isNew();
    }

    @Override
    public ID getId(@NotNull T entity) {
        return entity.getId();
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/repository/support/RequeryPersistableEntityInformation.kt
package org.springframework.data.requery.kotlin.repository.support

import io.requery.meta.EntityModel
import mu.KotlinLogging
import org.springframework.data.domain.Persistable
import kotlin.reflect.KClass

/**
 * org.springframework.data.requery.repository.support.RequeryPersistableEntityInformation
 *
 * @author debop
 */
class RequeryPersistableEntityInformation<E: Persistable<ID>, ID: Any>(kotlinType: KClass<E>,
                                                                       entityModel: EntityModel)
    : RequeryEntityModelEntityInformation<E, ID>(kotlinType, entityModel) {

    companion object {
        private val log = KotlinLogging.logger { }
    }

    init {
        log.debug { "Create RequeryPersistableEntityInformation. kotlinType=$kotlinType, entityModel=${entityModel.name}" }
    }

    override fun isNew(entity: E): Boolean {
        log.trace { "Is new entity?. entity=$entity" }
        return entity.isNew
    }

    override fun getId(entity: E): ID? {
        log.trace { "Get identifier of entity. entity=$entity" }
        return entity.id
    }

}